src/nubank/gen_cert.rs

```diff
@@ -0,0 +1,216 @@
+        .certificate;
+
+    let cert_bin = get_cert_bin(&cert_str, &key)?;
+
+    let full_path = format!("{}/{}", cert_folder, "cert.p12");
+    let _ = fs::write(&full_path, cert_bin).map_err(GenCertError::FailedToWriteCert)?;
+    Ok(full_path)
+}
+
+fn gen_private_key() -> Result<PKey<Private>, GenCertError> {
+    let rsa = Rsa::generate(2048).map_err(GenCertError::FailedToGeneratePrivateKey)?;
+    let private_key = PKey::from_rsa(rsa).map_err(GenCertError::FailedToGeneratePrivateKey);
+    private_key
+}
+
+fn build_payload(
+    cpf: &str,
+    password: &str,
+    key1: &PKey<Private>,
+    key2: &PKey<Private>,
+    device_id: String,
+) -> Result<HashMap<String, String>, GenCertError> {
+    let pub_key1 = get_public_key(key1)?;
+    let pub_key2 = get_public_key(key2)?;
+    let model = format!("PyNubank Client ({})", device_id);
+    let mut payload: HashMap<String, String> = HashMap::new();
+    payload.insert("login".to_string(), cpf.to_string());
+    payload.insert("password".to_string(), password.to_string());
+    payload.insert("public_key".to_string(), pub_key1);
+    payload.insert("public_key_crypto".to_string(), pub_key2);
+    payload.insert("model".to_string(), model);
+    payload.insert("device_id".to_string(), device_id);
+    Ok(payload)
+}
+
+fn build_new_payload(
+    old_payload: HashMap<String, String>,
+    encrypted_code: &str,
+    code: &str,
+) -> HashMap<String, String> {
+    let mut new_payload = old_payload.clone();
+    new_payload.insert("code".to_string(), code.to_string());
+    new_payload.insert("encrypted-code".to_string(), encrypted_code.to_string());
+    new_payload
+}
+
+fn get_public_key(key: &PKey<Private>) -> Result<String, GenCertError> {
+    let pub_key = key
+        .public_key_to_pem()
+        .map_err(GenCertError::FailedToGeneratePublicKey)?;
+    let pub_key_str = str::from_utf8(pub_key.as_slice())
+        .map_err(GenCertError::FailedToConvertPublicKeyToStr)?
+        .to_string();
+    Ok(pub_key_str)
+}
+
+fn parse_header_value(
+    header_value: &reqwest::header::HeaderValue,
+) -> Result<HashMap<String, String>, GenCertError> {
+    let mut parsed_header_value: HashMap<String, String> = HashMap::new();
+    let header_value_str = header_value
+        .to_str()
+        .map_err(GenCertError::FailedToConvertHeaderValueToStr)?;
+    let chunks = header_value_str.split(",");
+    for chunk in chunks {
+        let mut items = chunk.split("=");
+        let (key, value) = match items.clone().count() {
+            2 => (items.next().unwrap(), items.next().unwrap()),
+            _ => {
+                return Err(GenCertError::FailedToReadHeaderValue(
+                    header_value_str.to_string(),
+                ))
+            }
+        };
+        let parsed_key = key.trim().replace(" ", "_");
+        let parsed_value = value.replace(" ", "_");
+        parsed_header_value.insert(parsed_key, parsed_value);
+    }
```

Comment:
What about using iterators/recursion instead of a for loop? It would increase readability. For example: instead of using a for and a HashMap, you can use a normal Map and return it in a fold/reduce.

Reply:
I forgot that Rust doesn't have an immutable Map in the std lib. You can still use fold with HashMap btw.

Reply:
Done.